File: atcoder/abc302/d.cpp
#include<bits/stdc++.h>

using namespace std;
using ll = long long int;
const int maxn = 200001;

ll n, m, d;
vector<ll> cands_for_A;
vector<ll> cands_for_B;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n >> m >> d;
    for(int i = 0; i < n; i++) {
        ll x; cin >> x;
        cands_for_A.push_back(x);
    } 

    for (int i = 0; i < m; i++){
        ll x;
        cin >> x;
        cands_for_B.push_back(x);
    }
    sort(cands_for_A.begin(), cands_for_A.end());
    sort(cands_for_B.begin(), cands_for_B.end());

    // A - B <= D
    // 최대값으로 가려면 어떻게 해야돼?
    // B가 A와 동일한 값이어야 해
    ll ans = -1;
    for(int i = 0; i < n; i++) {
        ll A = cands_for_A[i];
        auto it = lower_bound(cands_for_B.begin(), cands_for_B.end(), A) - cands_for_B.begin();
        ll B = cands_for_B[it];

        // cout << A << " " << B << " " << it << endl;

        if(abs(A - B) <= d) {
            ans = max(ans, A + B);
        }
    }

    cout << ans << endl;

    return 0;
}

File: atcoder/abc302/d.py
import sys
from bisect import bisect_left, bisect_right
n, m, d = map(int, sys.stdin.readline().split())
cands_for_A = list(map(int, sys.stdin.readline().split()))
cands_for_B = list(map(int, sys.stdin.readline().split()))

cands_for_A.sort()
cands_for_B.sort()

ans = -1

for i in range(n):
    A = cands_for_A[i]
    it = bisect_left(cands_for_B, A + d)
    if it == m:
        it -= 1
    B = cands_for_B[it]
    if abs(A - B) <= d:
        ans = max(ans, A + B)

    if it + 1 < m:
        B = cands_for_B[it + 1]
        if abs(A - B) <= d:
            ans = max(ans, A + B)

for i in range(n):
    A = cands_for_A[i]
    it = bisect_right(cands_for_B, A - d)
    if it == m:
        it -= 1
    B = cands_for_B[it]
    if abs(A - B) <= d:
        ans = max(ans, A + B)

    if it + 1 < m:
        B = cands_for_B[it + 1]
        if abs(A - B) <= d:
            ans = max(ans, A + B)

for i in range(m):
    A = cands_for_B[i]
    it = bisect_left(cands_for_A, A + d)
    if it == n:
        it -= 1
    B = cands_for_A[it]
    if abs(A - B) <= d:
        ans = max(ans, A + B)

    if it + 1 < n:
        B = cands_for_A[it + 1]
        if abs(A - B) <= d:
            ans = max(ans, A + B)

for i in range(m):
    A = cands_for_B[i]
    it = bisect_right(cands_for_A, A - d)
    if it == n:
        it -= 1
    B = cands_for_A[it]
    if abs(A - B) <= d:
        ans = max(ans, A + B)

    if it + 1 < n:
        B = cands_for_A[it + 1]
        if abs(A - B) <= d:
            ans = max(ans, A + B)

print(ans)